### PR TITLE
Add RootSyncSet controller

### PR DESCRIPTION
**Pull request by mortent**
This adds the RootSyncSet controller to porch. This also copies code for token exchange to support workload identity from `pkg/registry/porch/wi` to `pkg/tokenexchange`.

This builds on https://github.com/GoogleContainerTools/kpt/pull/3456, so it must be merged after.

### CODE REVIEW

File: porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// RootSyncSetSpec defines the desired state of RootSyncSet
+type RootSyncSetSpec struct {
+	ClusterRefs []*ClusterRef `json:"clusterRefs,omitempty"`
```
Comment (johnbelamaric) [inline porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:23]: Embedding these directly doesn't work at scale. Are we sure we want to introduce the concept of "Cluster"?

File: porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	Template    *RootSyncInfo `json:"template,omitempty"`
```
Comment (johnbelamaric) [inline porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:24]: So we expect no variation between clusters?

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
I may be just out of the loop, but I am yet convinced on this concept at all. It's not clear to what problem it is solving. Can we get a use case description before we see code? I can't determine if this is the right approach before having a better understanding of the problem.

File: porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go (hunk 1 above)
Comment (justinsb) [reply porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:23]: I'd be fine with using "Target" instead (I think I'd personally prefer it), but I think we can iterate here.

File: porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go (hunk 2 above)
Comment (justinsb) [reply porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:24]: As I understand the plan, we're starting here and then going to figure out what variation we need.

File: porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go (hunk 1 above)
Comment (johnbelamaric) [reply porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:23]: I guess this controller uses direct API access so cluster makes sense

Comment (mortent) [reply porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:23]: Yeah, it possible that this doesn't work at large scale. We can iterate on this when we get some experience with this.

File: porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go (hunk 2 above)
Comment (mortent) [reply porch/controllers/rootsyncset/api/v1alpha1/rootsyncset_types.go:24]: Yeah, no variation is a starting point. We know that for some (maybe many or all) cases we do need more than this.

**Review by justinsb — APPROVED** [review]
Awesome - lgtm!

**Pull request merged**